projects/com.oracle.truffle.llvm.libraries.bitcode/src/syscall.c

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright (c) 2017, Oracle and/or its affiliates.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are
+ * permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this list of
+ * conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this list of
+ * conditions and the following disclaimer in the documentation and/or other materials provided
+ * with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to
+ * endorse or promote products derived from this software without specific prior written
+ * permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS
+ * OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdint.h>
+#include <stdarg.h>
+#include <errno.h>
+#include "syscall.h"
+
+__attribute__((weak)) int64_t syscall(int64_t n, ...) {
+  va_list ap;
+  int64_t a, b, c, d, e, f;
+  va_start(ap, n);
+  a = va_arg(ap, int64_t);
+  b = va_arg(ap, int64_t);
+  c = va_arg(ap, int64_t);
+  d = va_arg(ap, int64_t);
+  e = va_arg(ap, int64_t);
+  f = va_arg(ap, int64_t);
```

Comment:
AFAIK Sulong allocates a memory area for the GPRs, but only writes values there if they are available. `va_arg` reads from that area and may read garbage if nothing was written there. I will implement an intrinsic for the `syscall` function to avoid another problem: arguments are read as `int64_t` but might have been passed as something different (e.g. `short`). In this case, the value has to be sign extended to 32bit (that's the de-facto ABI on x86_64) and when a 32bit register is written, it is zero extended to 64bit. Since the call site doesn't know the prototype of the syscall, something like `syscall(something, -1)` might produce wrong results right now.